test/bmv2_parsing_check.bzl

```diff
@@ -0,0 +1,298 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+load("//:p4c.bzl", "run_p4c")
+
+def _parse_bmv2_impl(ctx):
+    # come up with a base name in a tmp directory
+    # for the .json and .pb.txt output files
+    basename = "".join([
+        ctx.attr.name,
+        "-parse-bmv2-tmp/",
+        ctx.file.input.basename[:-3]
+    ])
+
+    # declare output files
+    protobuf = ctx.actions.declare_file(
+        basename + ".pb.txt",
+        sibling=ctx.file.input
+    )
+    json = ctx.actions.declare_file(
+        basename + ".json",
+        sibling=ctx.file.input
+    )
+
+    # run the compiled binary to produce outputs
+    ctx.actions.run_shell(
+        inputs = [ctx.file._main, ctx.file.input],
+        outputs = [protobuf, json],
+        use_default_shell_env = True,
+        command = "{binary} {protobuf} {json} < {input}".format(
+            binary = ctx.file._main.path,
+            protobuf = protobuf.path,
+            json = json.path,
+            input = ctx.file.input.path
+        )
+    )
+
+    # return all outputs by default
+    # as well as a single output group per json and protobuf
+    return [
+        DefaultInfo(
+            files = depset([protobuf, json]),
+            runfiles = ctx.runfiles(files = [ctx.file.input, ctx.file._main])
+        ),
+        OutputGroupInfo(
+            json = depset([json]),
+            protobuf = depset([protobuf])
+        )
+    ]
+
+parse_bmv2 = rule(
+    doc = """"Runs our protobuf-based parsing binary on
+              an input bmv2 json (or a rule producing it),
+              and outputs a protobuf and json dump files.""",
+    implementation = _parse_bmv2_impl,
+    attrs = {
+        "input": attr.label(
+            doc = "Input json bmv2 program to feed to our parsing binary.",
+            mandatory = True,
+            allow_single_file = True
+        ),
+        "_main": attr.label(
+            doc = "src/main.cc binary that parses with protobuf.",
+            mandatory = False,
+            allow_single_file = True,
+            default = "//:main"
+        )
+    }
+)
+
+def _extract_output_group_impl(ctx):
+    output_group_name = ctx.attr.output_group
+    output_group = ctx.attr.target[OutputGroupInfo][output_group_name]
+    return [DefaultInfo(files = output_group)]
+
+extract_output_group = rule(
+    doc = """Extract the specified output group from a provided target..
+             Always targets to depend on the specified output group only,
```

Comment:
I'm not sure, but I think there is a typo in this sentence. Doesn't quite make sense right now, I think?